### PR TITLE
Complaints for modeling and text exercises

### PR DESCRIPTION
**Pull request by SiggZ**
### Checklist
- [x] I run `yarn run webpack:build:main`: the project builds without errors.
- [x] I run `yarn lint`: the project builds without code style warnings.
- [x] ~I updated the documentation and models.~
- [x] I tested the changes and all related features on the test server https://artemistest.ase.in.tum.de.
- [x] ~I added (end-to-end) test cases for the new functionality.~

### Description
In https://github.com/ls1intum/ArTEMiS/pull/353 we extended the existing complaints mechanism and integrated it into modeling exercises. In this PR we re-enable the complaints for text exercises and integrate the new changes. See https://github.com/ls1intum/ArTEMiS/pull/353 for further details.

### Steps for Testing
1. Log in to ArTEMiS as a student
2. Navigate to assessed text submission
3. Click on "Complain"
4. Enter and submit a complaint
5. Log in to ArTEMiS as a tutor (not the assessor of the submission from above)
6. Navigate to tutor exercise dashboard for the corresponding exercise
7. See list of complaints and click on "Evaluate the complaint"
8. Reply to complaint
9. Log in as student again
10. Check response for previously submitted complaint

### CODE REVIEW

**Comment by rpadovani** [comment]
I have tested both with text and modeling exercises, I have checked also the instructors dashboards and everything seems to work fine to me!

File: src/main/java/de/tum/in/www1/artemis/domain/AssessmentUpdate.java [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+/**
+ * A helper object that aggregates a feedback list and a complaint response which is used to update an assessment after
+ * a complaint.
+ */
+public class AssessmentUpdate {
+
+    /**
+     * The updated feedback list
+     */
+    private List<Feedback> feedbacks;
```
Comment (rpadovani) [inline src/main/java/de/tum/in/www1/artemis/domain/AssessmentUpdate.java]: `feedback` is unaccountable, please call this `feedbackList` (also in the rest of the code, and `getFeedbacks` become `getFeedbackList`)

File: src/main/java/de/tum/in/www1/artemis/repository/ComplaintRepository.java [hunk 2]
```diff
@@ -38,6 +41,17 @@
...
 
+    /**
+     * Count the number of unaccepted complaints of a student in a given course. Unaccepted means that they are either open/unhandled or rejected. We use this to limit the number
+     * of complaints for a student in a course.
+     *
+     * @param studentId the id of the student
+     * @param courseId  the id of the course
+     * @return the number of unaccepted
+     */
+    @Query("SELECT count(c) FROM Complaint c  WHERE c.student.id = :#{#studentId} AND c.result.participation.exercise.course.id = :#{#courseId} AND (c.accepted = false OR c.accepted is null)")
```
Comment (rpadovani) [inline src/main/java/de/tum/in/www1/artemis/repository/ComplaintRepository.java]: Does this always work? You don't do a `FETCH` :thinking:

File: src/main/java/de/tum/in/www1/artemis/domain/AssessmentUpdate.java (hunk 1 above)
Comment (SiggZ) [reply src/main/java/de/tum/in/www1/artemis/domain/AssessmentUpdate.java]: Well I know, but it's called `feedbacks` all over the code, so I tried to be somehow consistent. Of course I can change it, but I don't think that it makes much of a difference. Maybe we should discuss and agree on how to call the feedback list in general.

File: src/main/java/de/tum/in/www1/artemis/repository/ComplaintRepository.java (hunk 2 above)
Comment (SiggZ) [reply src/main/java/de/tum/in/www1/artemis/repository/ComplaintRepository.java]: Yes, I think it should work like this. We do not need a `FETCH` here as all the fields we are referring to are fetched eagerly from the database anyway. Or am I missing something?

File: src/main/webapp/app/complaints/complaints.component.ts [hunk 3]
```diff
@@ -15,30 +15,36 @@ import { ComplaintResponse } from 'app/entities/complaint-response';
...
     @Input() resultId: number;
+    @Input() allowedComplaints: number; // the number of complaints that a student can still submit in the course
     complaintText = '';
     alreadySubmitted: boolean;
     submittedDate: Moment;
     accepted: boolean;
+    handled: boolean;
     complaintResponse: ComplaintResponse;
 
+    readonly maxComplaintNumberPerStudent = 3;
```
Comment (krusche) [inline src/main/webapp/app/complaints/complaints.component.ts]:
```suggestion
    readonly maxComplaintNumberPerStudent = 3; // please note that this number has to be the same as in Constant.java on the server
```

**Pull request merged**